### DISCUSSION TITLE
Connections & RPSThreadGroup

### DISCUSSION THREAD

**Discussion by rabelenda** [abstracta/jmeter-java-dsl#86] [Q&A] [answered]
This comes from a exchange in discord which I think is worth having as general recipe or be aware:

How can we reuse connections between iterations using `rpsThreadGroup` and avoid it's dynamic handling of threads messing with connections handling?

**Comment by rabelenda** [comment, marked as answer]
Well there are 2 things to take into consideration:

* JMeter, by default, opens and closes iteration on each iteration. With standard JMeter Thread Group you can set either to use same user or not in each iteration, which allows reusing HTTP connections (if using HTTP 1.1 and keep-alives). But,  `rpsThreadGroup`, as mentioned in [user guide](https://abstracta.github.io/jmeter-java-dsl/guide/#throughput-based-thread-group), internally uses Concurrency Thread Group, which doesn't provide such option. So, to overcome this limitation the only option is to set the JMeter property `httpclient.reset_state_on_thread_group_iteration` to false.

* Even with previous setting, you have to take into consideration that HTTP connections are tied to Threads. In this regard `rpsThreadGroups` dynamically adjusts number of threads needed to satisfy a certain throughput requirement. This might create new threads when currently active threads are not enough for generating required load, or may on the other hand terminate threads when the number of threads is above the needed number to satisfy the load (TPS). Taking this into consideration, this means that connections could even be destroyed and re created as threads are destroyed and re created. To avoid this, you can use `spareThreads` method in `rpsThreadGroup` to always keep a buffer of threads and don't immediately terminate them when they are no longer needed. This setting is by default to `0.1` which means it will keep 10% of threads when they are no longer needed. So for example, if at some point is using 90 threads, and at a given moment it detects that needs 80 threads, it will not stop 10 threads, but only 2, leaving 88 threads (80 + 80*01). This would avoid unnecessarily closing threads, and in consequence connections, due to temporal changes in requirements of threads. You can increase this value to be more resilient to such changes. If you set the value to a number less than 1, it will treat it as a percentage, if you set it to 1 or greater, then it will use it as fixed number of threads to keep alive.

**In summary, if you use `rpsThreadGroup` and want to keep connections open, consider setting JMeter property `httpclient.reset_state_on_thread_group_iteration` to false and tuning `spareThreads()` setting.**
